pkg/mutation/system_external_data.go

```diff
@@ -0,0 +1,210 @@
+package mutation
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/open-policy-agent/frameworks/constraint/pkg/apis/externaldata/v1alpha1"
+	"github.com/open-policy-agent/frameworks/constraint/pkg/externaldata"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/types"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// resolvePlaceholders resolves all external data placeholders in the given object.
+func (s *System) resolvePlaceholders(obj *unstructured.Unstructured) error {
+	providerKeys := make(map[string]sets.String)
+	hasPlaceholders := false
+
+	// recurse object to find all existing external data placeholders
+	var recurse func(object interface{})
+	recurse = func(object interface{}) {
+		switch obj := object.(type) {
+		case map[string]interface{}:
+			// our base case - check if this object is an external data placeholder
+			if placeholder, ok := types.ToExternalDataPlaceholder(obj); ok {
+				hasPlaceholders = true
+				if _, ok := providerKeys[placeholder.ProviderName]; !ok {
+					providerKeys[placeholder.ProviderName] = sets.NewString()
+				}
+				// gather and de-duplicate all keys for this
+				// provider so we can resolve them in batch
+				providerKeys[placeholder.ProviderName].Insert(placeholder.ValueAtLocation)
+				return
+			}
+			for _, v := range obj {
+				recurse(v)
+			}
+		case []interface{}:
+			for _, item := range obj {
+				recurse(item)
+			}
+		}
+	}
+
+	recurse(obj.Object)
+	if !hasPlaceholders {
+		return nil
+	}
+
+	responses, err := s.sendRequests(providerKeys)
+	if err != nil {
+		return err
+	}
+
+	if err := s.mutateWithExternalData(obj, responses); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// sendRequests sends requests to all providers in parallel.
+func (s *System) sendRequests(providerKeys map[string]sets.String) (map[string]map[string]*externaldata.Item, error) {
+	responses := make(map[string]map[string]*externaldata.Item)
+	grp, _ := errgroup.WithContext(context.Background())
+	for name, keys := range providerKeys {
+		provider, err := s.providerCache.Get(name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get external data provider %s: %w", name, err)
```

Comment:
We'll probably need to record errors and then continue, so that we can honor the error handling specified by individual mutators (e.g. fall back to default value).

If we return immediately on error, then we may miss out on necessary return data.

The first non-nil error cancels the error group... instead preserving the error as a response object (to be handled later) may be helpful here.